src/storage/write_ahead_log/log_serializer_task.cpp

```diff
@@ -82,41 +84,66 @@ std::tuple<uint64_t, uint64_t, uint64_t> LogSerializerTask::Process() {
         std::unique_lock<std::mutex> guard(flush_queue_latch_);
 
         // There are no new buffers, so we can break
-        if (flush_queue_.empty()) {
+        if (disk_flush_queue_.empty() && replication_flush_queue_.empty()) {
           break;
         }
 
-        temp_flush_queue_ = std::move(flush_queue_);
-        flush_queue_ = std::queue<RecordBufferSegment *>();
+        temp_disk_flush_queue_ = std::move(disk_flush_queue_);
+        temp_replication_flush_queue_ = std::move(replication_flush_queue_);
+        disk_flush_queue_ = std::queue<RecordBufferSegment *>();
+        replication_flush_queue_ = std::queue<RecordBufferSegment *>();
         empty_ = true;
       }
 
       // Loop over all the new buffers we found
-      while (!temp_flush_queue_.empty()) {
-        RecordBufferSegment *buffer = temp_flush_queue_.front();
-        temp_flush_queue_.pop();
+      while (!temp_disk_flush_queue_.empty()) {
+        RecordBufferSegment *buffer = temp_disk_flush_queue_.front();
+        temp_disk_flush_queue_.pop();
 
         // Serialize the Redo buffer and release it to the buffer pool
         IterableBufferSegment<LogRecord> task_buffer(buffer);
-        const auto num_bytes_records_and_txns = SerializeBuffer(&task_buffer);
+        const auto num_bytes_records_and_txns = SerializeBuffer(&task_buffer, SerializeDestination::DISK);
         buffer_pool_->Release(buffer);
         num_bytes += std::get<0>(num_bytes_records_and_txns);
         num_records += std::get<1>(num_bytes_records_and_txns);
         num_txns += std::get<2>(num_bytes_records_and_txns);
       }
 
-      buffers_processed = true;
+      disk_buffers_processed = true;
+
+      if (!temp_replication_flush_queue_.empty()) {
+        replication_buffers_processed = true;
+      }
+
+      while (!temp_replication_flush_queue_.empty()) {
+        RecordBufferSegment *buffer = temp_replication_flush_queue_.front();
+        temp_replication_flush_queue_.pop();
+
+        // Serialize the Redo buffer and release it to the buffer pool
+        IterableBufferSegment<LogRecord> task_buffer(buffer);
+        const auto num_bytes_records_and_txns = SerializeBuffer(&task_buffer, SerializeDestination::REPLICAS);
+        buffer_pool_->Release(buffer);
+
+        num_bytes += std::get<0>(num_bytes_records_and_txns);
+        num_records += std::get<1>(num_bytes_records_and_txns);
+        num_txns += std::get<2>(num_bytes_records_and_txns);
+      }
     }
 
     // Mark the last buffer that was written to as full
-    if (buffers_processed) HandFilledBufferToWriter();
+    if (disk_buffers_processed) HandFilledBufferToWriter(SerializeDestination::DISK);
```

Comment:
I think there's a couple of logic errors with these boolean flags. Consider the following two examples

1. `disk_flush_queue` and `replication_flush_queue` both have log records.
2. The loop on line 99 completes, fills `disk_filled_buffer_` with logs, and marks `disk_buffers_processed` as true.
3. `temp_replication_flush_queue_` is not empty so the `if` on line 114 evaluates to false
4. The loop on line 118 completes, fills `replication_filled_buffer_` with logs, and DOES NOT mark `replication_buffers_processed` true.
5. The loop repeats `disk_flush_queue_.empty() && replication_flush_queue_.empty()` evaluates to true and we break out of the loop.
Now even though we have serialized log records for the replica `replication_buffers_processed` is false and we do not continue processing them.

Even if `disk_flush_queue_` is empty, we ALWAYS set `disk_buffers_processed` to true. I don't think this is correct. We end up trying to send logs to the disk log consumer even though we didn't serialize any.